vulcanai2/models/basenetwork.py

```diff
@@ -551,15 +592,8 @@ def _validate(self, val_loader):
 
         for batch_idx, (data, targets) in enumerate(val_loader):
 
-            # for idx, d in enumerate(data):
-            #     data[idx] = Variable(d, requires_grad=False)
-            # targets = Variable(targets, requires_grad=False)
-
-            if torch.cuda.is_available():
```

Comment:
@priyatharsan don't you need to also move the optimizer to the specified device? or is that implicitly done.

Reply:
Yeah it is done implicitly, I just checked it. You can get the device by loop through the optimizer's state_dict, something like: 
```
for state in optimizer.state.values():
    for k, v in state.items():
        if torch.is_tensor(v):
            #state[k] = v.is_cuda
            state[k] = v.device
```

Reply:
there might be issues when switching devices in between fit sessions from what I read online. try to exhaustively test all the variations and create unit tests for what you can

Reply:
wait! we are not switching device between fit sessions, do we? also, can you share the resource you are referring to?

Reply:
What Im referring to is being able to train on gpu and then train on cpu if needed, and vice versa. This issue explains it clearly: https://github.com/pytorch/pytorch/issues/2830

Reply:
I think the best way to handle it is just to rebuild the optimizer instead of distributing the param_groups to the devices of the respective networks?

Reply:
but it's not ideal since you forget the state_dict of the optimizer which isn't ideal for continuing to train

Reply:
That is the recommended way so far. But we could try experimenting with the optimi.param.device